lib/ai/chat_test.go

```diff
@@ -0,0 +1,214 @@
+
+package ai
+
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/sashabaranov/go-openai"
+	"github.com/stretchr/testify/require"
+	"github.com/tiktoken-go/tokenizer/codec"
+)
+
+func TestChat_PromptTokens(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name     string
+		messages []openai.ChatCompletionMessage
+		want     int
+		wantErr  bool
+	}{
+		{
+			name:     "empty",
+			messages: []openai.ChatCompletionMessage{},
+			want:     3,
+		},
+		{
+			name: "only system message",
+			messages: []openai.ChatCompletionMessage{
+				{
+					Role:    openai.ChatMessageRoleSystem,
+					Content: "Hello",
+				},
+			},
+			want: 8,
+		},
+		{
+			name: "system and user messages",
+			messages: []openai.ChatCompletionMessage{
+				{
+					Role:    openai.ChatMessageRoleSystem,
+					Content: "Hello",
+				},
+				{
+					Role:    openai.ChatMessageRoleUser,
+					Content: "Hi LLM.",
+				},
+			},
+			want: 16,
+		},
+		{
+			name: "tokenize our prompt",
+			messages: []openai.ChatCompletionMessage{
+				{
+					Role:    openai.ChatMessageRoleSystem,
+					Content: promptCharacter("Bob"),
+				},
+				{
+					Role:    openai.ChatMessageRoleUser,
+					Content: "Show me free disk space on localhost node.",
+				},
+			},
+			want: 187,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			chat := &Chat{
+				messages:  tt.messages,
+				tokenizer: codec.NewCl100kBase(),
+			}
+			usedTokens, err := chat.PromptTokens()
+			require.NoError(t, err)
+			require.Equal(t, tt.want, usedTokens)
+		})
+	}
+}
+
+func TestChat_Complete(t *testing.T) {
+	t.Parallel()
+
+	responses := [][]byte{
+		generateTextResponse(),
+		generateCommandResponse(),
+	}
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "text/event-stream")
+
+		require.GreaterOrEqual(t, len(responses), 1, "Unexpected request")
+		dataBytes := responses[0]
+
+		_, err := w.Write(dataBytes)
+		require.NoError(t, err, "Write error")
+
+		responses = responses[1:]
+	}))
+	defer server.Close()
+
+	cfg := openai.DefaultConfig("secret-test-token")
+	cfg.BaseURL = server.URL + "/v1"
+	client := NewClientFromConfig(cfg)
+
+	chat := client.NewChat("Bob")
+
+	t.Run("initial message", func(t *testing.T) {
+		msg, tokens, err := chat.Complete(context.Background())
+		require.NoError(t, err)
+
+		expectedResp := &Message{Role: "assistant",
+			Content: "Hey, I'm Teleport - a powerful tool that can assist you in managing your Teleport cluster via ChatGPT.",
+			Idx:     0,
+		}
+		require.Equal(t, expectedResp, msg)
+		require.Equal(t, 0, tokens)
+	})
+
+	t.Run("text completion", func(t *testing.T) {
+		chat.Insert(openai.ChatMessageRoleUser, "Show me free disk space")
+
+		msg, tokens, err := chat.Complete(context.Background())
+		require.NoError(t, err)
+		// TODO(justinas): Here we should get 4 tokens, but the API reports 1.
```

Comment:
@justinas Is this a bug or I'm using the API in the wrong way?

Reply:
So with `StreamingMessage`, the tokens flow asynchronously (via a channel) to the caller of `Complete`, so `numTokens` is kind of useless.

`processComplete()` in the dev-ai branch already handles this correctly, but I'll modify the signature to make it less confusing.

Reply:
See https://github.com/gravitational/teleport/pull/26058